libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,579 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * Stack layout generator for Yul to EVM code generation.
+ */
+
+#include <libyul/backends/evm/StackLayoutGenerator.h>
+
+#include <libyul/backends/evm/StackHelpers.h>
+
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/cxx20.h>
+#include <libsolutil/Visitor.h>
+
+#include <range/v3/algorithm/any_of.hpp>
+#include <range/v3/range/conversion.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/concat.hpp>
+#include <range/v3/view/drop.hpp>
+#include <range/v3/view/drop_last.hpp>
+#include <range/v3/view/filter.hpp>
+#include <range/v3/view/iota.hpp>
+#include <range/v3/view/map.hpp>
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/take.hpp>
+#include <range/v3/view/transform.hpp>
+
+using namespace solidity;
+using namespace solidity::yul;
+using namespace std;
+
+StackLayout StackLayoutGenerator::run(CFG const& _cfg)
+{
+	StackLayout stackLayout;
+	StackLayoutGenerator{stackLayout, {}}.processEntryPoint(*_cfg.entry);
+
+	for (auto& functionInfo: _cfg.functionInfo | ranges::views::values)
+		StackLayoutGenerator{stackLayout, functionInfo.returnVariables}.processEntryPoint(*functionInfo.entry);
+
+	return stackLayout;
+}
+
+StackLayoutGenerator::StackLayoutGenerator(StackLayout& _layout, vector<VariableSlot> _currentFunctionReturnVariables):
+	m_layout(_layout),
+	m_currentFunctionReturnVariables(move(_currentFunctionReturnVariables))
+{
+}
+
+namespace
+{
+
+/// @returns the ideal stack to have before executing an operation that outputs @a _operationOutput, s.t.
+/// shuffling to @a _post is cheap.
+/// If @a _generateSlotOnTheFly returns true for a slot, this slot should not occur in the ideal stack, but
+/// rather be generated on the fly during shuffling.
+template<typename Callable>
+Stack createIdealLayout(Stack const& _operationOutput, Stack const& _post, Callable _generateSlotOnTheFly)
+{
+	struct PreviousSlot { size_t slot; };
+
+	// Determine the number of slots that have to be on stack before executing the operation.
+	// That is slots that should not be generated on the fly and are not outputs of the operation.
+	size_t previousLayoutSize = _post.size();
+	for (auto const& slot: _post)
+		if (util::findOffset(_operationOutput, slot) || _generateSlotOnTheFly(slot))
```

Comment:
Is this
```suggestion
		if (contains(_operationOutput, slot) || _generateSlotOnTheFly(slot))
```
?

Reply:
Yeah... I might be using it like that on a few occasions, I'll try to change all uses of ``findOffset`` that can be replaced by ``contains`` with ``contains``.